ibis/pandas/execution.py

```diff
@@ -0,0 +1,438 @@
+def execute_cast_series_date(op, data, type, scope=None):
+    return data.dt.date
+
+
+_LITERAL_CAST_TYPES = {
+    dt.double: float,
+    dt.float: float,
+    dt.int64: int,
+    dt.int32: int,
+    dt.int16: int,
+    dt.int8: int,
+    dt.string: str,
+    dt.timestamp: pd.Timestamp,
+}
+
+
+@execute_node.register(ops.Cast, scalar_types, dt.DataType)
+def execute_cast_string_literal(op, data, type, scope=None):
+    try:
+        return _LITERAL_CAST_TYPES[type](data)
+    except KeyError:
+        raise TypeError(
+            "Don't know how to cast {!r} to type {}".format(data, type)
+        )
+
+
+@execute_node.register(ops.TableColumn, pd.DataFrame)
+def execute_table_column_dataframe(op, data, scope=None):
+    return data[op.name]
+
+
+@execute_node.register(ops.Selection, pd.DataFrame)
+def execute_selection_dataframe(op, data, scope=None):
+    selections = op.selections
+    predicates = op.predicates
+    sort_keys = op.sort_keys
+
+    result = data
+
+    if selections:
+        old_names = [s.op().name for s in selections]
+        new_names = [
+            getattr(s, '_name', old_name) or old_name
+            for old_name, s in zip(old_names, selections)
+        ]
+        result = result[old_names].rename(dict(zip(old_names, new_names)))
+
+    if predicates:
+        where = functools.reduce(
+            operator.and_, (execute(p, scope) for p in predicates)
+        )
+        result = result.loc[where]
+
+    if sort_keys:
+        return result.sort_values(sort_keys)
+    return result
+
+
+def reduction_to_lambda(expr):
+    op = expr.op()
+
+    def reduction(
+        df,
+        column=op.args[0]._name,
+        method_name=type(op).__name__.lower(),
+        new_name=expr._name,
+    ):
+        method = getattr(df[column], method_name)
+        return method().rename(new_name)
+    return reduction
+
+
+@execute_node.register(ops.Aggregation, pd.DataFrame)
+def execute_aggregation_dataframe(op, data, scope=None):
+    assert op.agg_exprs
+    assert op.by
+
+    if op.having:
+        raise NotImplementedError('having expressions not yet implemented')
+
+    if op.sort_keys:
+        raise NotImplementedError(
+            'sorting on aggregations not yet implemented'
+        )
+
+    predicates = op.predicates
+    if predicates:
+        data = data.loc[
+            functools.reduce(
+                operator.and_, (execute(p, scope) for p in predicates)
+            )
+        ]
+
+    gb = data.groupby([execute(by, scope) for by in op.by])
+
+    aggs = [reduction_to_lambda(agg_expr) for agg_expr in op.agg_exprs]
+    pieces = [func(gb) for func in aggs]
+    agg_df = pd.concat(pieces, axis=1)
+
+    return agg_df.reset_index()
+
+
+@execute_node.register(ops.Reduction, pd.Series, (pd.Series, type(None)))
+def execute_reduction_series_mask(op, data, mask, scope=None):
+    operand = data[mask] if mask is not None else data
+    return getattr(operand, type(op).__name__.lower())()
```

Comment:
this is good as a pattern, since we can implement faster reductions on filtered data in libpandas